### PR TITLE
Use ResourceManager for GetAllStrings

### PR DESCRIPTION
**Pull request by ryanbrandenburg**
`AssemblyResourceStringProvider` was originally created as a workaround for a missing API in `ResourceManager`, now that that API is back let's use it.

### CODE REVIEW

**Comment by hishamco** [comment]
@ryanbrandenburg can we merge both your PR & mine #384 to ensure that we fixed the workaround permanently?

File: src/Microsoft.Extensions.Localization/Internal/ResourceManagerStringProvider.cs [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+            return _resourceNamesCache.GetOrAdd(cacheKey, _ =>
+            {
+                var resourceSet = _resourceManager.GetResourceSet(culture, true, false);
+
+                if(resourceSet == null)
+                {
+                    return ThrowOrNull(culture, throwOnMissing);
+                }
+
+
```
Comment (hishamco) [inline src/Microsoft.Extensions.Localization/Internal/ResourceManagerStringProvider.cs]: Remove the extra line here

File: test/Microsoft.Extensions.Localization.Tests/ResourceManagerStringLocalizerTest.cs [hunk 2]
```diff
@@ -233,49 +242,57 @@ private static int GetCultureInfoDepth(CultureInfo culture)
...
-            public int GetManifestResourceStreamCallCount { get; private set; }
+            public TestAssemblyWrapper(Type type)
+                : base(type.GetTypeInfo().Assembly)
+            {
+            }
 
-            public override string FullName { get; }
+            public bool HasResources { get; set; } = true;
+
+            public int GetManifestResourceStreamCallCount { get; private set; }
```
Comment (hishamco) [inline test/Microsoft.Extensions.Localization.Tests/ResourceManagerStringLocalizerTest.cs]: `ManifestResourceStreamCallCount` is better than `GetManifestResourceStreamCallCount`

File: src/Microsoft.Extensions.Localization/Internal/ResourceManagerStringProvider.cs [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+        {
+            var cacheKey = GetResourceCacheKey(culture);
+
+            return _resourceNamesCache.GetOrAdd(cacheKey, _ =>
+            {
+                var resourceSet = _resourceManager.GetResourceSet(culture, true, false);
+
+                if(resourceSet == null)
+                {
+                    return ThrowOrNull(culture, throwOnMissing);
```
Comment (pranavkm) [inline src/Microsoft.Extensions.Localization/Internal/ResourceManagerStringProvider.cs]: Inline this. No good reason for this to be a separate method.

File: src/Microsoft.Extensions.Localization/Internal/ResourceManagerStringProvider.cs [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+            return null;
+        }
+
+        public IList<string> GetAllResourceStrings(CultureInfo culture, bool throwOnMissing)
+        {
+            var cacheKey = GetResourceCacheKey(culture);
+
+            return _resourceNamesCache.GetOrAdd(cacheKey, _ =>
+            {
+                var resourceSet = _resourceManager.GetResourceSet(culture, true, false);
```
Comment (pranavkm) [inline src/Microsoft.Extensions.Localization/Internal/ResourceManagerStringProvider.cs]: named params.

File: src/Microsoft.Extensions.Localization/Internal/ResourceManagerStringProvider.cs [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+                if(resourceSet == null)
+                {
+                    return ThrowOrNull(culture, throwOnMissing);
+                }
+
+
+                var rs = resourceSet.GetEnumerator();
```
Comment (pranavkm) [inline src/Microsoft.Extensions.Localization/Internal/ResourceManagerStringProvider.cs]: This is unused.

File: src/Microsoft.Extensions.Localization/Internal/ResourceManagerStringProvider.cs (hunk 4 above)
Comment (pranavkm) [reply src/Microsoft.Extensions.Localization/Internal/ResourceManagerStringProvider.cs]: `resourceSet` needs to be disposed.

File: src/Microsoft.Extensions.Localization/Internal/ResourceManagerStringProvider.cs [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+
+                var names = new List<string>();
+                foreach(DictionaryEntry i in resourceSet)
```
Comment (pranavkm) [inline src/Microsoft.Extensions.Localization/Internal/ResourceManagerStringProvider.cs]: space after `foreach`

Comment (pranavkm) [reply src/Microsoft.Extensions.Localization/Internal/ResourceManagerStringProvider.cs]: `DictonaryEntry entry`

File: src/Microsoft.Extensions.Localization/Internal/ResourceManagerStringProvider.cs [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+                {
+                    names.Add(i.Key.ToString());
```
Comment (pranavkm) [inline src/Microsoft.Extensions.Localization/Internal/ResourceManagerStringProvider.cs]: `names.Add((string)entry.Key);`

File: src/Microsoft.Extensions.Localization/Internal/ResourceManagerStringProvider.cs [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+
+        public IList<string> GetAllResourceStrings(CultureInfo culture, bool throwOnMissing)
+        {
+            var cacheKey = GetResourceCacheKey(culture);
+
+            return _resourceNamesCache.GetOrAdd(cacheKey, _ =>
+            {
+                var resourceSet = _resourceManager.GetResourceSet(culture, true, false);
+
+                if(resourceSet == null)
```
Comment (pranavkm) [inline src/Microsoft.Extensions.Localization/Internal/ResourceManagerStringProvider.cs]: formatting.

File: test/Microsoft.Extensions.Localization.Tests/ResourceManagerStringLocalizerTest.cs [hunk 9]
```diff
@@ -11,6 +11,7 @@
 using Microsoft.Extensions.Logging;
 using Microsoft.Extensions.Logging.Testing;
 using Xunit;
+using System;
```
Comment (pranavkm) [inline test/Microsoft.Extensions.Localization.Tests/ResourceManagerStringLocalizerTest.cs]: sort usings

**Comment by ryanbrandenburg** [comment]
@hishamco it's unclear to me what your PR solves that mine doesn't, could you elaborate?

File: src/Microsoft.Extensions.Localization/Internal/ResourceManagerStringProvider.cs (hunk 4 above)
Comment (ryanbrandenburg) [reply src/Microsoft.Extensions.Localization/Internal/ResourceManagerStringProvider.cs]: When I put `ResourceSet` in a using I get [this](https://ci.appveyor.com/project/aspnetci/localization-r5928/build/1.0.930#L436) when we do GetString later on. As discussing in person, won't fix.

**Comment by ryanbrandenburg** [comment]
:up::date:

**Comment by pranavkm** [comment]
@hishamco Isn't switching over to `GetResourceSet` sufficient to remove the workaround we had for the missing API? What does #384 do in addition?

**Comment by hishamco** [comment]
> it's unclear to me what your PR solves that mine doesn't, could you elaborate?

I thought your PR to fix the workaround permanently, because  I already start the that ..

> Isn't switching over to GetResourceSet sufficient to remove the workaround we had for the missing API?

AFAIK the workaround is done for the missed API, Am I right?

**Comment by Eilon** [comment]
@hishamco sorry we're not quite sure what you mean. Is there a particular issue with this PR? It seems to address some issues with the original PR you had sent out, so we think it's closer to being something we can merge into the 2.1 release and avoid breaking changes.

**Comment by hishamco** [comment]
No problem, the important thing we close the bugs 😄

**Pull request merged**